examples/06-multiple-connections/server.c

```diff
@@ -0,0 +1,455 @@
+	ret = epoll_add(svr->epoll, &clnt->ev_conn_event);
+	if (ret)
+		goto err_sync_disconnect;
+
+	/* get the connection's completion file descriptor */
+	ret = rpma_conn_get_completion_fd(clnt->conn, &fd);
+	if (ret) {
+		print_error_ex("rpma_conn_get_completion_fd", ret);
+		(void) rpma_conn_disconnect(clnt->conn);
+		return;
+	}
+
+	/* set O_NONBLOCK flag for the connection's completion fd */
+	ret = fd_set_nonblock(fd);
+	if (ret) {
+		(void) rpma_conn_disconnect(clnt->conn);
+		return;
+	}
+
+	/* prepare a custom event structure and add it to the epoll */
+	custom_event_set(&clnt->ev_conn_cmpl, fd, clnt, conn_cmpl);
+	ret = epoll_add(svr->epoll, &clnt->ev_conn_cmpl);
+	if (ret)
+		(void) rpma_conn_disconnect(clnt->conn);
+
+	return;
+
+err_sync_disconnect:
+	(void) common_disconnect_and_wait_for_conn_close(&clnt->conn);
+	--svr->client_num;
+}
+
+int
+main(int argc, char *argv[])
+{
+	/* validate parameters */
+	if (argc < 3) {
+		fprintf(stderr, USAGE_STR, argv[0]);
+		exit(-1);
+	}
+
+	/* read common parameters */
+	char *addr = argv[1];
+	char *service = argv[2];
+	int ret;
+
+	/* RPMA resources - general */
+	struct rpma_peer *peer = NULL;
+
+	/* server resource */
+	struct server_res svr = {0};
+
+	/* scalable I/O */
+	int ep_fd;
+	svr.epoll = epoll_create1(EPOLL_CLOEXEC);
+	if (svr.epoll == -1)
+		return errno;
+
+	/*
+	 * lookup an ibv_context via the address and create a new peer using it
+	 */
+	ret = server_peer_via_address(addr, &peer);
+	if (ret)
+		goto err_close_epoll;
+
+	/* allocate a memory */
+	size_t dst_size = DWORD * CLIENT_MAX;
+	svr.dst_ptr = malloc_aligned(dst_size);
+	if (svr.dst_ptr == NULL) {
+		ret = -1;
+		goto err_peer_delete;
+	}
+
+	/* register the memory */
+	ret = rpma_mr_reg(peer, svr.dst_ptr, dst_size, RPMA_MR_USAGE_READ_DST,
+			RPMA_MR_PLT_VOLATILE, &svr.dst_mr);
+	if (ret) {
+		print_error_ex("rpma_mr_reg", ret);
+		goto err_mr_free;
+	}
+
+	/* start a listening endpoint at addr:service */
+	ret = server_listen(peer, addr, service, &svr.ep);
+	if (ret)
+		goto err_mr_dereg;
+
+	/* get the endpoint's event file descriptor */
+	ret = rpma_ep_get_fd(svr.ep, &ep_fd);
+	if (ret) {
+		print_error_ex("rpma_ep_get_fd", ret);
+		goto err_ep_shutdown;
+	}
+
+	/* set O_NONBLOCK flag for the endpoint's event fd */
+	ret = fd_set_nonblock(ep_fd);
+	if (ret)
+		goto err_ep_shutdown;
+
+	/* prepare a custom event structure and add it to the epoll */
+	custom_event_set(&svr.ev_incoming, ep_fd, &svr, ep_incoming_func);
+	ret = epoll_add(svr.epoll, &svr.ev_incoming);
+	if (ret)
+		goto err_ep_shutdown;
+
+	(void) printf("Today I have got in touch with:\n");
+
+	/* process epoll's events */
+	struct epoll_event event;
+	struct custom_event *ce;
+	do {
+		ret = epoll_wait(svr.epoll, &event, 1 /* # of events */,
+				TIMEOUT_15s);
+		if (ret == -1)
+			break;
+
+		if (ret == 1) {
+			ce = (struct custom_event *)event.data.ptr;
+			ce->func(ce);
+		} else if (ret == 0) { /* timeout */
```

Comment:
sooo, `while ((ret = epoll_wait) == 1)` ?

I usually prefer to use loops that are as simple as possible in terms of control flow.